src/ruby/puppet-server-lib/puppet/server/network/http/handler.rb

```diff
@@ -32,7 +34,12 @@ def path(request)
   end
 
   def body(request)
-    request["body"]
+    body = request["body"]
+    if body.java_kind_of?(InputStream)
+      body.to_io.read()
```

Comment:
Correction, IO#read does not do an in-memory copy if I understand http://ruby-doc.org/stdlib-1.9.3/libdoc/stringio/rdoc/StringIO.html#method-i-read